### ISSUE TITLE
upstreamForceH2C could be independent with insecureListenAddress

### ISSUE THREAD

**Issue by andylibrian** [brancz/kube-rbac-proxy#139]
Hi, Thanks for the work on this proxy. I have a question, is there any reason `--upstream-force-h2c` is tied to `--insecure-listen-address`?

https://github.com/brancz/kube-rbac-proxy/blob/fa09bd61bfd7fd1b6312aac131a1f65e5cf01296/main.go#L355-L356

My use case is that the upstream is a plain / insecure gRPC, but the kube-rbac-proxy runs with TLS (as a sidecar). I tried to move this code block (`if cfg.upstreamForceH2C {` ) out of `insecureListenAddress` and it worked for my case.

If you agree to untie these config, I will send a PR.

**Comment by s-urbaniak** [comment]
@andylibrian per definition h2c (http/2 over TCP) is cleartext and not encrypted, so combining it with a secure upstream transport does not suit.

**Comment by s-urbaniak** [comment]
@andylibrian scratch my comment, i see what you mean :thinking: indeed, you are right. having a secure terminating endpoint with kube-rbac-proxy while having an insecure upstream actually does make sense.

Happy to review the PR once you have it.

**Comment by s-urbaniak** [comment]
I am not sure if reverseproxy plays well with a h2c insecure upstream, see https://github.com/brancz/kube-rbac-proxy/pull/2/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R198.

**Comment by andylibrian** [comment]
@s-urbaniak Thanks for taking a second look, and for the information about h2c.

Looking at the referenced issue in the comment you mentioned (https://github.com/golang/go/issues/14141#issuecomment-388594444), it seems like h2c has been implemented in `golang.org/x/net/http2/h2c`. Then I saw this commit (26399541337f92eea845d32ffa3afd7596099188) updated the h2c implementation from `github.com/hkwi/h2c` to `golang.org/x/net/http2/h2c` . So, It's probably working well now.
